libgalois/include/katana/PropertyIndex.h

```diff
@@ -0,0 +1,192 @@
+#ifndef KATANA_LIBGALOIS_KATANA_PROPERTYINDEX_H_
+#define KATANA_LIBGALOIS_KATANA_PROPERTYINDEX_H_
+
+#include <set>
+#include <string>
+#include <variant>
+
+#include <arrow/api.h>
+#include <arrow/array.h>
+#include <arrow/type_traits.h>
+#include <boost/iterator/iterator_categories.hpp>
+#include <boost/iterator/iterator_facade.hpp>
+
+#include "katana/Result.h"
+#include "katana/config.h"
+
+namespace katana {
+
+// PropertyIndex provides an interface similar to an ordered container
+// over a single property.
+template <typename node_or_edge>
+class KATANA_EXPORT PropertyIndex {
+public:
+  // Type-safe container for node and edge ids to avoid overlap with the
+  // primitive types we are indexing.
+  struct IndexId {
+    node_or_edge id;
+  };
+
+  // The set key type is either a node/edge id (all the keys in the actual set)
+  // or a value representing the search key.
+  using set_key_type = typename std::variant<
+      IndexId, int32_t, int64_t, double_t, std::string_view*>;
+
+  // PropertyIndex::iterator returns a sequence of node or edge ids.
+  using base_iterator = typename std::multiset<set_key_type>::iterator;
+  class iterator : public base_iterator {
+  public:
+    explicit iterator(const base_iterator& base) : base_iterator(base) {}
+    const node_or_edge& operator*() const {
+      return std::get<IndexId>(base_iterator::operator*()).id;
+    }
+  };
+
+  PropertyIndex(std::string column) : column_(std::move(column)) {}
+
+  // The name of the indexed property.
+  std::string column() { return column_; }
+
+  virtual iterator begin() = 0;
+  virtual iterator end() = 0;
+
+  virtual Result<void> BuildFromProperty() = 0;
+  // virtual Result<void> BuildFromFile() = 0;
+
+private:
+  std::string column_;
+};
+
+// PrimitivePropertyIndex provides a PropertyIndex for primitive types.
+template <typename node_or_edge, typename c_type>
+class KATANA_EXPORT PrimitivePropertyIndex
+    : public PropertyIndex<node_or_edge> {
+public:
+  using ArrowArrayType = typename arrow::CTypeTraits<c_type>::ArrayType;
+  using IndexId = typename PropertyIndex<node_or_edge>::IndexId;
+  using iterator = typename PropertyIndex<node_or_edge>::iterator;
+  using set_key_type = typename PropertyIndex<node_or_edge>::set_key_type;
+
+  PrimitivePropertyIndex(
+      const std::string& column, size_t num_entities,
+      std::shared_ptr<arrow::Array> property)
+      : PropertyIndex<node_or_edge>(column),
+        num_entities_(num_entities),
```

Comment:
Both are true - I named this num_entities because it's meant to be the number of nodes/edges. This also happens to be the number of rows in the table/array, but it wasn't clear that that was an invariant.